src/Opaleye/Internal/HaskellDB/PrimQuery.hs

```diff
@@ -60,6 +60,7 @@ data BinOp      = (:==) | (:<) | (:<=) | (:>) | (:>=) | (:<>)
                 | (:->) | (:->>) | (:#>) | (:#>>)
                 | (:@>) | (:<@) | (:?) | (:?|) | (:?&)
                 | (:&&) | (:<<) | (:>>) | (:&<) | (:&>) | (:-|-)
+                | OpArrayIndex
```

Comment:
I definitely want this as a new constructor.  If you want to open a new issue or PR to discuss the organisation of that module then please go ahead.  Such a discussion doesn't belong in this PR though.